python/python/tests/test_file.py

```diff
@@ -189,3 +195,53 @@ def test_list_field_name(tmp_path):
 
     assert round_tripped == table
     assert round_tripped.schema.field("list_str").type == weird_string_type
+
+
+def test_dictionary(tmp_path):
+    # Basic round trip
+    dictionary = pa.array(["foo", "bar", "baz"], pa.string())
+    indices = pa.array([0, 1, 2, 0, 1, 2], pa.int32())
+    dict_arr = pa.DictionaryArray.from_arrays(indices, dictionary)
+
+    def round_trip(arr):
+        table = pa.table({"dict": arr})
+
+        path = tmp_path / "foo.lance"
+        with LanceFileWriter(str(path)) as writer:
+            writer.write_batch(table)
+
+        reader = LanceFileReader(str(path))
+        table2 = reader.read_all().to_table()
+        return table2.column("dict").chunk(0)
+
+    round_tripped = round_trip(dict_arr)
+
+    assert round_tripped == dict_arr
+    assert round_tripped.type == dict_arr.type
+
+    # Dictionary that doesn't use all values
+    dictionary = pa.array(["foo", "bar", "baz"], pa.string())
+    indices = pa.array([0, 0, 1, 1], pa.int32())
+    dict_arr = pa.DictionaryArray.from_arrays(indices, dictionary)
+
+    round_tripped = round_trip(dict_arr)
+
+    assert round_tripped.dictionary == dictionary
```

Comment:
Is this the desired behavior? I'm kind of surprised we wouldn't remove unused values (or at least allow that in the future).

Reply:
If the user is giving us dictionary encoded data then the dictionary may be meaningful to them and I didn't want to lose that information.  I agree it would be nice to have an option to vacuum dictionaries on write.

Alternatively, we may want to make a distinction between "enum" and "dictionary" data types which is something [polars](https://docs.pola.rs/user-guide/concepts/data-types/categoricals/#enum-vs-categorical) does.  Enum is an "extension type" (with storage type dictionary)  In the enum case you would store the dictionary at the column level and do not vacuum it.  In the categorical case you would always vacuum dictionaries.